File: clojure/bob/src/bob.clj
(ns bob
  (:use clojure.string))

(defn question? [s]
  (ends-with? s "?"))

(defn all-caps? [s]
  (and
    (re-find #"[A-Z]" s)
    (= s (upper-case s))))

(defn all-caps-question? [s]
  (and (all-caps? s) (question? s)))

(defn response-for [s] ;; <- arglist goes here
  (cond
    (blank? s) "Fine. Be that way!"
    (all-caps-question? s) "Calm down, I know what I'm doing!"
    (question? s) "Sure."
    (all-caps? s) "Whoa, chill out!"
    :else "Whatever."))

File: python/bob/bob.py
import re

def hey(phrase):
    stripped = phrase.strip()
    if not stripped:
        return "Fine. Be that way!"

    question = __is_question(stripped)
    shouting = __is_shouting(stripped)

    if question and shouting:
        return "Calm down, I know what I'm doing!"
    elif question:
        return "Sure."
    elif shouting:
        return "Whoa, chill out!"
    else:
        return "Whatever."

def __is_shouting(phrase):
    return (bool(re.match(r'[A-Z\s\W\d]+\Z', phrase)) and
            bool(re.search(r'[A-Z]', phrase)))

def __is_question(phrase):
    return bool(re.match(r'.*\?\Z', phrase))
